### PR TITLE
Use docker compose to bring up the application for Acceptance Tests

### PR DESCRIPTION
**Pull request by chrisgilmerproj**
## Description

This is an (almost) working example of using docker-compose to bring up the application from a specific git commit (we can write machinery to get that commit from a branch name later). I would like to consider this as an alternative to bringing up AWS infra.

## Setup

Setup is fairly straight forward if you have `brew` installed and are running on OSX:

```sh
# This first step not needed to test this PR but is what you'd do in future
git checkout <branch_name> 

# Here are the rest of the steps you'll do every time
direnv allow
make docker_compose_setup
make docker_compose_up
make docker_compose_down
```

You can use these to log in:

-   http://milmovelocal:5000/devlocal-auth/login
-  http://officelocal:5000/devlocal-auth/login
- http://tsplocal:5000/devlocal-auth/login
- http://adminlocal:5000/devlocal-auth/login

This is a complete working server and everything is self contained. It relies on github/circleci building and pushing your image to ECR (which is private, but we could either build a proxy to make them public or use dockerhub or wherever we decide). 

This avoids folks needing any dependencies that developers need to set up an environment (go, yarn, pre-commit, etc ... basically anything in `scripts/prereqs`) so its fairly light weight and portable. Plus, the docker layers are cached so startup takes seconds. 

For reference: AWS deploys would be 20-30 minutes plus all the extra maintenance burden managing the infra would require. So using docker-compose a PM could very quickly switch between branches and test them before signing off on pivotal tasks vs waiting potentially hours between tests.

### CODE REVIEW

**Comment by chrisgilmerproj** [comment]
> Having issues:
> 
> ```
> milmove_1          | 2019-07-29T22:25:32.174Z   ERROR   internalapi/users.go:42 Error retrieving servic
> e_member        {"git_branch": "cg_grant_all_ecs_user", "git_commit": "2a0262233d7bd7632694fd09f2eedcfc
> de8a80b1", "milmove_trace_id": "87bd14b5-3ecf-4d51-ae4e-efbadf1a21ec", "error": "dial tcp 172.18.0.2:54
> 36: connect: connection refused"}
> ```
> 
> I think this may be related to not being able to connect to the db. Need to do further investigation.

So the migrations don't run because unlike the app container the migration container expects the DB to be up before it comes up. The app can start without a DB and can just wait idly for it.  Probably we just need the same wait logic in the migration code. The issue is really that docker-compose can't know when postgres is ready to accept connections in order to hold off on bringing up the migrations. This is totally solvable fwiw I just didn't have time to fix it.

I should note that the app comes up as expected and all the logs for that look normal.

**Comment by chrisgilmerproj** [comment]
At this point the only issue is the DB connection between docker containers. Once the containers are started with `docker-compose up` you can test that the docker container is created with `dev_db` outside of docker compose using this:

```sh
cd scripts
go run github.com/transcom/mymove/cmd/milmove migrate --db-host localhost --db-port 6432 --db-name dev_db --db-user postgres --db-ssl-mode disable -m ../migrations_manifest.txt -p ../migrations
```

**NOTE:** That command will fail on the first secure migration but that's expected.

I suggest commenting out the `milmove` container and just focusing on the `milmove_migrations` container connection. I think with this hurdle crossed we'll have a good working example.

File: docker-compose.yml [hunk 1]
```diff
@@ -0,0 +1,84 @@
...
+       - DB_ENV=development
+       - DB_HOST=database
+       - DB_NAME=dev_db
+       - DB_PASSWORD=mysecretpassword
+       - DB_PORT=5432
+       - DB_SSL_MODE=disable
+       - DB_USER=postgres
+       - ENVIRONMENT=test
+       - SECURE_MIGRATION_DIR=/mymove/local_migrations
+       - SECURE_MIGRATION_SOURCE
```
Comment (chrisgilmerproj) [inline docker-compose.yml]: We should set this to `local` explicitly.

File: docker-compose.yml [hunk 2]
```diff
@@ -0,0 +1,84 @@
...
+       - LOGIN_GOV_SECRET_KEY
+       - LOGIN_GOV_TSP_CLIENT_ID
+       - LOGIN_GOV_ADMIN_CLIENT_ID
+       - MOVE_MIL_DOD_CA_CERT
+       - MOVE_MIL_DOD_TLS_CERT
+       - MOVE_MIL_DOD_TLS_KEY
+       - NO_TLS_ENABLED=1
+       - NO_TLS_PORT=5000
+       - SECURE_MIGRATION_DIR
+       - SECURE_MIGRATION_SOURCE
```
Comment (chrisgilmerproj) [inline docker-compose.yml]: I don't think we need `SECURE*` for the app container.

File: scripts/update-docker-compose [hunk 3]
```diff
@@ -0,0 +1,7 @@
+#! /usr/bin/env bash
+
+set -eux -o pipefail
```
Comment (chrisgilmerproj) [inline scripts/update-docker-compose]: Need to remove the `x` here. We only needed it for debugging.

File: docker-compose.yml [hunk 4]
```diff
@@ -0,0 +1,84 @@
...
+     ports:
+       - '5000:5000'
+     environment:
+       - CLIENT_AUTH_SECRET_KEY
+       - CSRF_AUTH_KEY
+       - DB_ENV=development
+       - DB_HOST=database
+       - DB_NAME=dev_db
+       - DB_PASSWORD=mysecretpassword
+       - DB_PORT=6432
```
Comment (chrisgilmerproj) [inline docker-compose.yml]: Set this to `5432`

File: Makefile [hunk 5]
```diff
@@ -923,4 +923,27 @@ spellcheck: .client_deps.stamp # Run interactive spellchecker
...
+.PHONY: docker_compose_setup
+docker_compose_setup: ## Install requirements to use docker-compose
+	brew install git docker docker-compose direnv aws-vault
+
+.PHONY: docker_compose_up
+docker_compose_up: ## Bring up docker-compose containers
+	aws ecr get-login --no-include-email --region us-west-2 --no-include-email | sh
+	scripts/update-docker-compose
+	@echo "Open app at http://milmovelocal:5000"
+	docker-compose up
```
Comment (chrisgilmerproj) [inline Makefile]: Should we launch this with `-d` so that the logs don't print?

File: cmd/milmove/migrate.go [hunk 6]
```diff
@@ -147,15 +148,22 @@ func migrateFunction(cmd *cobra.Command, args []string) error {
...
+		dbConnection, errDbConn = cli.InitDatabase(v, logger)
+		if errDbConn != nil {
+			if dbConnection == nil {
+				// No connection object means that the configuraton failed to validate and we should kill server startup
+				logger.Fatal("Invalid DB Configuration", zap.Error(errDbConn))
+			} else {
+				// A valid connection object that still has an error indicates that the DB is not up and
+				// thus is not ready for migrations
+				logger.Error("DB is not ready for connections, sleeping", zap.Error(errDbConn))
+				time.Sleep(5 * time.Second)
```
Comment (chrisgilmerproj) [inline cmd/milmove/migrate.go]: I'm not comfortable letting this indefinitely cycle. We probably should have this break after a certain number of tries (like 10?).

**Comment by chrisgilmerproj** [comment]
What is committed at this point in the branch runs the entire app. It's worth testing for anyone connected to this branch. While there's probably still some refinement this is a great start.

**Comment by chrisgilmerproj** [comment]
> Issues encountered:
> 
>     * `make docker_compose_install` is actually `make docker_compose_setup` (probably just an issue with the instructions on this PR.)

Fixed this in the instructions.

> 
>     * `make docker_compose_setup` errors for me (I probably already have the deps installed):
> 
> 
> ```
> ...
> Error: No available formula with the name "aws-vault" 
> Found a cask named "aws-vault" instead.
> Did you mean to type "brew cask install aws-vault"?
> make: *** [docker_compose_setup] Error 1
> ```

You're right, you probably already have these installed. But we can make it not error.

>     * `make docker_compose_up` fails for product_manager role:
> 
> 
> ```
> $ make docker_compose_up
> aws ecr get-login --no-include-email --region us-west-2 --no-include-email | sh
> 
> An error occurred (AccessDeniedException) when calling the GetAuthorizationToken operation: User: arn:aws:sts::923914045601:assumed-role/product-manager/1564518723574536000 is not authorized to perform: ecr:GetAuthorizationToken on resource: *
> scripts/update-docker-compose
> Open app at http://milmovelocal:5000
> docker-compose up
> Creating network "mymove_default" with the default driver
> Pulling milmove_migrate (923914045601.dkr.ecr.us-west-2.amazonaws.com/app-migrations:git-branch-cg_docker_compose)...
> ERROR: Get https://923914045601.dkr.ecr.us-west-2.amazonaws.com/v2/app-migrations/manifests/git-branch-cg_docker_compose: no basic auth credentials
> make: *** [docker_compose_up] Error 1
> ```

You need [AmazonEC2ContainerRegistryReadOnly](https://docs.aws.amazon.com/AmazonECR/latest/userguide/ecr_managed_policies.html#AmazonEC2ContainerRegistryReadOnly) permissions which I'll add now.

**Comment by chrisgilmerproj** [comment]
@akostibas - try to pull down images now that we've applied https://github.com/transcom/ppp-infra/pull/578

File: Makefile (hunk 5 above)
Comment (chrisgilmerproj) [reply Makefile]: Team says no for this: concern is things will be running forever in the background if its not visible.

File: cmd/milmove/migrate.go (hunk 6 above)
Comment (chrisgilmerproj) [reply cmd/milmove/migrate.go]: Revert this and try using docker-compose health_check.

File: cmd/milmove/migrate.go [hunk 7]
```diff
@@ -146,16 +147,32 @@ func migrateFunction(cmd *cobra.Command, args []string) error {
...
-	if err != nil {
-		if dbConnection == nil {
-			// No connection object means that the configuraton failed to validate and we should kill server startup
-			logger.Fatal("Invalid DB Configuration", zap.Error(err))
+	// Create a connection to the DB with retry logic
+	var dbConnection *pop.Connection
+	var errDbConn error
+	retry := 0
+	retryMax := 5
+	retryInterval := 5 * time.Second
```
Comment (chrisgilmerproj) [inline cmd/milmove/migrate.go]: These could be command line arguments. The trick is figuring out if we want to use this retry logic in other places or not.

**Comment by chrisgilmerproj** [comment]
We're calling this done until we get more feedback about requirements from PMs or an indication that something doesn't work. We don't want to drop too much time into this just yet.

PMs will need Github access and AWS accounts to use this tool but its fairly trivial to get them set up and we can do that anytime.

**Comment by akostibas** [comment]
Got it working! Very cool. It's faster than starting the services locally.

Once I connected to `milmovelocal:5000` (had to manually determine the port by looking at `docker ps`), ran into some issues:
* `milmovelocal:5000`: No local signin link is offered. Presumably the app doesn't know it's in a container running locally.
* `milmovelocal:5000`: Tried to click the login link, nothing seemingly happens.
* `officelocal:5000`: Tried login link, Login.gov doesn't like where the request originated from (`    Redirect uri redirect_uri does not match registered redirect_uri`)

**Comment by chrisgilmerproj** [comment]
> Got it working! Very cool. It's faster than starting the services locally.

Glad it worked!

> Once I connected to `milmovelocal:5000` (had to manually determine the port by looking at `docker ps`), ran into some issues:

We had an `echo` statement with this information but it got completely lost in the output from the `docker-compose` CLI so it was removed. Probably we just need a `how-to` doc where we list it.
 
>     * `milmovelocal:5000`: No local signin link is offered. Presumably the app doesn't know it's in a container running locally.

You have to manually browse to http://milmovelocal:5000/devlocal-auth/login to get what you want. Unfortunately the devlocal sign in button is blocked in production builds. This would be the same on actual AWS infra as well since we explicitly do not allow building development JS bundles for deployment to AWS.

>     * `milmovelocal:5000`: Tried to click the login link, nothing seemingly happens.

Likely its not configured to use login.gov correctly. Probably an env var somewhere ...

>     * `officelocal:5000`: Tried login link, Login.gov doesn't like where the request originated from (`    Redirect uri redirect_uri does not match registered redirect_uri`)

You can use these to log in:

- http://milmovelocal:5000/devlocal-auth/login
- http://officelocal:5000/devlocal-auth/login
- http://tsplocal:5000/devlocal-auth/login
- http://adminlocal:5000/devlocal-auth/login

> Once local signing gets working, I _think_ that this will have parity with running the service locally.

Local signing in works fine. And so does local user creation. If you're fancy you can build the `bin/generate-test-data` and run it against port 6432 and import all your e2e data (we don't have a good solution for building this yet but we can tackle it if its needed ... probably as another docker container).

As for using `login.gov` we probably need to figure that out but its probably not blocking for most of the stories that want to be reviewed.

**Comment by chrisgilmerproj** [comment]
> The local login interface is a little gross (needing to know the exact URLs), but it's a pretty nice setup for so little work. Basically: I think it's a good proof of concept that we can decide to invest in more.

You're right, having to know the URL stinks - but honestly there's not a good way to approach that without possibly compromising what we deploy to production. I bet if we just leave those in a "how-to" doc so people can click on them or if folks can bookmark them it will be fine.
 
> It would be nice if there was a way to test that the compose file actually builds. I could imagine all the env vars getting out of sync or something.

We pass through almost all the env vars that matter via the `.envrc`. The ones that need to not change or this thing will break are hard coded. If we decide the PMs don't want to use `direnv` then we'll have to hard code the remaining set of env vars and THAT will be an issue for falling out of sync and I'd also want a test then.

> 🚢

Thanks for the approval! I'm going to try this on app-experimental because of the migration file changes. And then I'll get this merged.  

PS - Writing this PR has given me a thought about how we might speed up integration tests. I may be able to shave off 4-ish minutes if I use prebuilt images. That's something to look into for the future.

**Comment by chrisgilmerproj** [comment]
This worked in experimental!

**Pull request merged**